### DISCUSSION TITLE
Lesson 11: Using VRF Coordinator

### DISCUSSION THREAD

**Discussion by alexkalen** [smartcontractkit/full-blockchain-solidity-course-py#1680] [Q&A] [answered]
How does the contract know that the specific token has not been already taken by another user?
In other words, how does the contract know what NFTs have been taken before asigning it to a new user order in the minting process?

**Comment by cromewar** [comment, marked as answer]
Hello @alexkalen 
Well it depends on the contract, the Openzeppelin ones does check that with a function called `safe_mint`, which verifies exactly that in order to prevent it. I highly recommend you to give a check to the Openzeppelin docs in order to understand deeply how they prevent this problem.

**Reply by alexkalen** [reply]
Alright. Thank you very much.

**Reply by alexkalen** [reply]
@cromewar one question you might know the answer to. Why do the chainlink docs say the fee for using VRF Coordinator is 0.1 LINK but then take out 1 whole LINK?

**Reply by cromewar** [reply]
Well, would like to know more context on the question but let's try to answer it.
The Fee is actually 0.1 LINK on the video Patrick send 1 entire Link token just for confort because is not actually needed, especially for VRF v1.
However based on the JS version of this course, VRF is now version 2 and it changed the method to obtain the random number a lot I've seen some users on that forum stating that if you don't have enough LINK (let's say 5 o more) the transaction will get reverted, even if the Fee is actually just 0.1.
The problem was solved just adding more funds, but is not clear if it's a bug, or their require you you have more LINK just to ensure you will be able to pay the next tx.

Anyways, as the course in Python is using v1, in this case Patrick just sent 1 for confort as I said before.

**Reply by alexkalen** [reply]
Ok that makes sense, thank you.

**Reply by alexkalen** [reply]
@cromewar Im back for a follow up regarding the first question I had. I checked the safe_mint docs and understood your point. I now understand you cant actually mint a token twice. But my question is regarding the mapping.  For example, in the code below, if the number 2 is generated twice with the module operation then the mapping would pair two tokenIds with the same identifier. Im not sure I understand how @PatrickAlphaC  avoids this. Especially when minting thousands of NFTs. 

``` 
 function fulfillRandomness(bytes32 requestId, uint256 randomNumber)
        internal
        override
    {
        identifier = (randomNumber % 3) + 1; // '3' es maxima capacidad de tokens
        uint256 newTokenId = tokenCounter;
        tokenIdToIdentifier[newTokenId] = identifier;
        emit NFTAssigned(newTokenId, identifier);
        address owner = requestIdToSender[requestId];
        _safeMint(owner, newTokenId);
        // _setTokenURI(newTokenId, tokenURI);
        tokenCounter = tokenCounter + 1;
    }
```

**Reply by alexkalen** [reply]
PD: I changed the whole Breed mapping because I wanted to have numbers instead. Thank you in advanced for helping me out.
